x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/AuthorizationService.java

```diff
@@ -171,30 +168,42 @@ public AuthorizationService(
     }
 
     public void checkPrivileges(
-        Authentication authentication,
-        HasPrivilegesRequest request,
+        Subject subject,
+        AuthorizationEngine.PrivilegesToCheck privilegesToCheck,
         Collection<ApplicationPrivilegeDescriptor> applicationPrivilegeDescriptors,
-        ActionListener<HasPrivilegesResponse> listener
+        ActionListener<AuthorizationEngine.PrivilegesCheckResult> listener
     ) {
-        getAuthorizationEngine(authentication).checkPrivileges(
-            authentication,
-            getAuthorizationInfoFromContext(),
-            request,
-            applicationPrivilegeDescriptors,
-            wrapPreservingContext(listener, threadContext)
+        final AuthorizationEngine authorizationEngine = getAuthorizationEngineForSubject(subject);
+        authorizationEngine.resolveAuthorizationInfo(
+            subject,
+            wrapPreservingContext(
+                listener.delegateFailure(
+                    (delegateListener, authorizationInfo) -> authorizationEngine.checkPrivileges(
+                        authorizationInfo,
+                        privilegesToCheck,
+                        applicationPrivilegeDescriptors,
+                        wrapPreservingContext(delegateListener, threadContext)
+                    )
+                ),
+                threadContext
+            )
         );
     }
 
-    public void retrieveUserPrivileges(
-        Authentication authentication,
-        GetUserPrivilegesRequest request,
-        ActionListener<GetUserPrivilegesResponse> listener
-    ) {
-        getAuthorizationEngine(authentication).getUserPrivileges(authentication, getAuthorizationInfoFromContext(), request, listener);
-    }
-
-    private AuthorizationInfo getAuthorizationInfoFromContext() {
-        return Objects.requireNonNull(threadContext.getTransient(AUTHORIZATION_INFO_KEY), "authorization info is missing from context");
+    public void retrieveUserPrivileges(Subject subject, ActionListener<GetUserPrivilegesResponse> listener) {
+        final AuthorizationEngine authorizationEngine = getAuthorizationEngineForSubject(subject);
+        authorizationEngine.resolveAuthorizationInfo(
+            subject,
+            wrapPreservingContext(
+                listener.delegateFailure(
+                    (delegateListener, authorizationInfo) -> authorizationEngine.getUserPrivileges(
+                        authorizationInfo,
+                        wrapPreservingContext(listener, threadContext)
+                    )
+                ),
+                threadContext
+            )
+        );
```

Comment:
The existing code gets `authorizationInfo` from the `threadContext` instead of computing it again. I think we should maintain this behaviour. It might be necessary to compute it if we support retrieving privileges by profileId in future. But we can update the code here when it is needed.

Reply:
I have pushed d9b60fbf63e865a09a3d87fcdbaad1c72b which reverts to the before behavior that uses the authz info from the context rather than "resolving" it again from the subject.